### PR TITLE
feat: context logger

### PR DESCRIPTION
**Pull request by SimonRichardson**
Introduces `context.Context` to the `logger.Logger` interface. The main driver behind this is the following:

- For all model facade requests, a model uuid is inserted into the context[1]. This will allow us to annotate all lot messages with a model uuid. This will pay dividends when we're exporting all logs to loki or similar. This also has the advantage that logs can be bifurcated based on the uuid to different log files for machines.
- Trace IDs can also be associated with log messages, so it will be possible to see which logs are associated with a trace.
- Status history epic. Long term we don't want to store the information in the database, instead we want the information be part of the logging system. For this to work, we need the model UUID to be part of the logs (first point), but more work is required to make sure the model UUID is part of the model workers. That will have to be added later.
- This opens up the avenue for using the std library slog package in the future.

1. https://github.com/juju/juju/blob/5e972995681ae74a95bef063a31e4e253e6157c7/apiserver/apiserver.go#L1085-L1088

---

The code is relatively simple, it just introduces a context.Context as part of the method signature and the wire up of the context either via a TODO or Background context is used.

## QA Steps

```sh
$ juju bootstrap lxd test
$ juju add-model default
$ juju deploy ubuntu
```

### CODE REVIEW

**Comment by manadart** [comment]
Got a couple of failing unit/smoke tests.

**Comment by hpidcock** [comment]
`This opens up the avenue for using the std library slog package in the future.` :100:

**Review by hpidcock — APPROVED** [review]
Wow, yeah, what a riveting read.

File: api/watcher/watcher.go [hunk 1]
```diff
@@ -368,6 +371,10 @@ func (w *relationUnitsWatcher) Changes() watcher.RelationUnitsChannel {
 	return w.out
 }
 
+func (w *relationUnitsWatcher) scopedContext() (context.Context, context.CancelFunc) {
```
[the remark this answers is not in the document]
Comment (SimonRichardson) [reply api/watcher/watcher.go:374]: It's going that way.

File: apiserver/apiserver.go [hunk 2]
```diff
@@ -730,6 +731,7 @@ func (srv *Server) endpoints() ([]apihttp.Endpoint, error) {
 	healthHandler := http.HandlerFunc(srv.healthHandler)
 	embeddedCLIHandler := newEmbeddedCLIHandler(httpCtxt)
 	debugLogHandler := newDebugLogTailerHandler(
+		context.TODO(),
```
[the remark this answers is not in the document]
Comment (SimonRichardson) [reply apiserver/apiserver.go:734]: Maybe it did at one point, I'll rewind it back.

**Comment by SimonRichardson** [comment]
Trying a less aggressive approach https://github.com/juju/juju/pull/18364

**Comment by tlm** [comment]
Are we going to close this in favour of ##18364 ?

**Pull request closed without merging**